src/common/ceph_crypto.h

```diff
@@ -20,61 +21,148 @@
 // with error checking, and just say these really should never fail.
 // This assert MUST NOT be compiled out, even on non-debug builds.
 # include "include/ceph_assert.h"
+#endif /*USE_NSS*/
+
+#ifdef USE_OPENSSL
+typedef struct env_md_ctx_st EVP_MD_CTX;
+typedef struct env_md_st EVP_MD;
+#endif /*USE_OPENSSL*/
 
 namespace ceph {
   namespace crypto {
     void assert_init();
     void init(CephContext *cct);
     void shutdown(bool shared=true);
     class Digest {
-    private:
-      PK11Context *ctx;
-      size_t digest_size;
     public:
-      Digest (SECOidTag _type, size_t _digest_size) : digest_size(_digest_size) {
-	ctx = PK11_CreateDigestContext(_type);
-	ceph_assert_always(ctx);
-	Restart();
-      }
-      ~Digest () {
-	PK11_DestroyContext(ctx, PR_TRUE);
-      }
-      void Restart() {
-	SECStatus s;
-	s = PK11_DigestBegin(ctx);
-	ceph_assert_always(s == SECSuccess);
-      }
-      void Update (const unsigned char *input, size_t length) {
-        if (length) {
+      Digest () {};
+      virtual ~Digest () {}
+      virtual void Restart() = 0;
+      virtual void Update (const unsigned char *input, size_t length) = 0;
+      virtual void Final (unsigned char *digest) = 0;
```

Comment:
i'd love to avoid calling through a virtual interface, and it doesn't look like this Digest interface really needs to exist

Reply:
Have removed.